packages/server/src/api/app/bsky/util/search.ts

```diff
@@ -0,0 +1,158 @@
+  const emptyQb = db.db
+    .selectFrom('user')
+    .where(sql`1 = 0`)
+    .select([sql.literal('').as('did'), sql<number>`0`.as('distance')])
+  const resultsQb = db.db
+    .selectFrom(
+      emptyQb
+        .union(sql`${accountsQb}`) // The sql`` is adding parens
+        .union(sql`${profilesQb}`)
+        .as('accounts_and_profiles'),
+    )
+    .selectAll()
+    .distinctOn('did') // Per did, take whichever of account and profile distance is best
+    .orderBy('did')
+    .orderBy('distance')
+
+  // Sort and paginate all user results
+  const keysetAll = keyset(cursor, {
+    username: ref('username'),
+    distance: ref('distance'),
+  })
+  return db.db
+    .selectFrom(resultsQb.as('results'))
+    .innerJoin('user', 'user.did', 'results.did')
+    .if(!!keysetAll, (qb) => (keysetAll ? qb.where(keysetAll) : qb))
+    .orderBy('distance')
+    .orderBy('username') // Keep order stable: break ties in distance arbitrarily using username
+    .limit(limit)
+}
+
+export const getUserSearchQuerySqlite = (
+  db: Database,
+  opts: { term: string; limit: number; before?: string },
+) => {
+  const { ref } = db.db.dynamic
+  const { term, limit, before } = opts
+
+  // Take the first three words in the search term. We're going to build a dynamic query
+  // based on the number of words, so to keep things predictable just ignore words 4 and
+  // beyond. We also remove the special wildcard characters supported by the LIKE operator,
+  // since that's where these values are heading.
+  const safeWords = term
+    .replace(/[%_]/g, '')
+    .split(/\s+/)
+    .filter(Boolean)
+    .slice(0, 3)
+
+  if (!safeWords.length) {
+    // Return no results. This could happen with weird input like ' % _ '.
+    return db.db.selectFrom('user').where(sql`1 = 0`)
+  }
+
+  // We'll ensured there's a space before each word in both textForMatch and in safeWords,
+  // so that we can reliably match word prefixes using LIKE operator.
+  const textForMatch = sql`lower(' ' || ${ref(
+    'user.username',
+  )} || ' ' || coalesce(${ref('profile.displayName')}, ''))`
+
+  const cursor = before !== undefined ? unpackCursor(before) : undefined
+
+  return db.db
+    .selectFrom('user')
+    .where((q) => {
+      safeWords.forEach((word) => {
+        // Match word prefixes against contents of username and displayName
+        q = q.where(textForMatch, 'like', `% ${word.toLowerCase()}%`)
+      })
+      return q
+    })
+    .if(!!cursor, (qb) =>
+      cursor ? qb.where('username', '>', cursor.name) : qb,
+    )
+    .orderBy('username')
+    .limit(limit)
+}
+
+// E.g. { distance: .94827, name: 'pfrazee' } -> '[0.94827,"pfrazee"]'
+export const packCursor = (row: { distance: number; name: string }): string => {
+  const { distance, name } = row
+  return JSON.stringify([distance, name])
+}
+
+export const unpackCursor = (
+  before: string,
+): { distance: number; name: string } => {
+  const result = JSON.parse(before) /// @TODO bourne
```

Comment:
what does this TODO mean?

is this a Jason Bourne joke? 😆

Reply:
Ah thanks for calling this out, I nearly forgot about it.  I think at the end of the day it actually is a Jason Bourne joke haha!  I was thinking of this library which I usually use for JSON parsing untrusted inputs, which can lead to proto poisoning https://github.com/hapijs/bourne.  This one is equally good too: https://github.com/fastify/secure-json-parse.

Because of the way the object is being used I know we aren't subject to proto poisoning, but it can be a nice practice to just use something that handles this so that you don't have to really think about it.

Reply:
I pulled bourne in for the parsing, also helped handle more types of bad cursors (e.g. if it's not JSON at all).